File: SWIFT/Packages/SwiftCore/Sources/Domain/TimetableDiffEngine.swift
import Foundation

public struct TimetableDiffEngine: Sendable {
    public init() {}

    /// Compares oldSnapshot vs newSnapshot and produces classified change records.
    public func diff(
        oldSnapshot: TimetableSnapshot?,
        newSnapshot: TimetableSnapshot
    ) -> [TimetableChangeRecord] {
        guard let oldSnapshot = oldSnapshot, !oldSnapshot.periods.isEmpty else {
            return []
        }

        // If academic year or week number changed, this is a new schedule import, not in-place adjustments
        if let oldWeek = oldSnapshot.weekNumber, let newWeek = newSnapshot.weekNumber, oldWeek != newWeek {
            return []
        }

        var changes: [TimetableChangeRecord] = []

        let oldDict: [String: PeriodOccurrence] = Dictionary(
            oldSnapshot.periods.map { ($0.id, $0) },
            uniquingKeysWith: { first, _ in first }
        )
        let newDict: [String: PeriodOccurrence] = Dictionary(
            newSnapshot.periods.map { ($0.id, $0) },
            uniquingKeysWith: { first, _ in first }
        )

        // 1. Detect updates and additions
        for newPeriod in newSnapshot.periods {
            if let oldPeriod = oldDict[newPeriod.id] {
                // Subject changed
                if newPeriod.subject != oldPeriod.subject {
                    changes.append(
                        TimetableChangeRecord(
                            periodId: newPeriod.id,
                            day: newPeriod.day,
                            dateString: newPeriod.dateString,
                            periodName: newPeriod.periodName,
                            changeType: .subjectChanged,
                            oldSubject: oldPeriod.subject,
                            newSubject: newPeriod.subject,
                            oldStartTime: oldPeriod.startTime,
                            newStartTime: newPeriod.startTime,
                            oldEndTime: oldPeriod.endTime,
                            newEndTime: newPeriod.endTime,
                            oldDetails: oldPeriod.details,
                            newDetails: newPeriod.details
                        )
                    )
                }

                // Time changed
                if newPeriod.startTime != oldPeriod.startTime || newPeriod.endTime != oldPeriod.endTime {
                    changes.append(
                        TimetableChangeRecord(
                            periodId: newPeriod.id,
                            day: newPeriod.day,
                            dateString: newPeriod.dateString,
                            periodName: newPeriod.periodName,
                            changeType: .timeChanged,
                            oldSubject: oldPeriod.subject,
                            newSubject: newPeriod.subject,
                            oldStartTime: oldPeriod.startTime,
                            newStartTime: newPeriod.startTime,
                            oldEndTime: oldPeriod.endTime,
                            newEndTime: newPeriod.endTime,
                            oldDetails: oldPeriod.details,
                            newDetails: newPeriod.details
                        )
                    )
                }

                // Details changed (only if subject didn't also change)
                if newPeriod.subject == oldPeriod.subject && newPeriod.details != oldPeriod.details && !newPeriod.details.isEmpty {
                    changes.append(
                        TimetableChangeRecord(
                            periodId: newPeriod.id,
                            day: newPeriod.day,
                            dateString: newPeriod.dateString,
                            periodName: newPeriod.periodName,
                            changeType: .detailsChanged,
                            oldSubject: oldPeriod.subject,
                            newSubject: newPeriod.subject,
                            oldStartTime: oldPeriod.startTime,
                            newStartTime: newPeriod.startTime,
                            oldEndTime: oldPeriod.endTime,
                            newEndTime: newPeriod.endTime,
                            oldDetails: oldPeriod.details,
                            newDetails: newPeriod.details
                        )
                    )
                }
            } else {
                // Newly added period slot
                changes.append(
                    TimetableChangeRecord(
                        periodId: newPeriod.id,
                        day: newPeriod.day,
                        dateString: newPeriod.dateString,
                        periodName: newPeriod.periodName,
                        changeType: .added,
                        newSubject: newPeriod.subject,
                        newStartTime: newPeriod.startTime,
                        newEndTime: newPeriod.endTime,
                        newDetails: newPeriod.details
                    )
                )
            }
        }

        // 2. Detect removed periods
        for oldPeriod in oldSnapshot.periods {
            if newDict[oldPeriod.id] == nil {
                changes.append(
                    TimetableChangeRecord(
                        periodId: oldPeriod.id,
                        day: oldPeriod.day,
                        dateString: oldPeriod.dateString,
                        periodName: oldPeriod.periodName,
                        changeType: .removed,
                        oldSubject: oldPeriod.subject,
                        oldStartTime: oldPeriod.startTime,
                        oldEndTime: oldPeriod.endTime,
                        oldDetails: oldPeriod.details
                    )
                )
            }
        }

        // If a massive number of periods changed (e.g. initial baseline import or format migration),
        // do not treat the whole timetable as changes.
        if changes.count > 15 {
            return []
        }

        return changes
    }

    /// Attaches change records to the matching periods in a snapshot for instant UI rendering.
    public func annotateSnapshot(
        _ snapshot: TimetableSnapshot,
        with changes: [TimetableChangeRecord]
    ) -> TimetableSnapshot {
        let changeMap: [String: TimetableChangeRecord] = Dictionary(
            changes.filter { !$0.isAcknowledged }.map { ($0.periodId, $0) },
            uniquingKeysWith: { current, _ in current }
        )

        let annotatedPeriods = snapshot.periods.map { period -> PeriodOccurrence in
            var p = period
            p.changeRecord = changeMap[period.id]
            return p
        }

        return TimetableSnapshot(
            academicYear: snapshot.academicYear,
            weekNumber: snapshot.weekNumber,
            startDate: snapshot.startDate,
            endDate: snapshot.endDate,
            periods: annotatedPeriods
        )
    }
}

File: src/Jadwal.Domain/Rules/TimetableDiffEngine.cs
using Jadwal.Domain.Enums;
using Jadwal.Domain.Models;

namespace Jadwal.Domain.Rules;

public class TimetableDiffEngine
{
    public IReadOnlyList<TimetableChangeRecord> Diff(TimetableSnapshot? oldSnapshot, TimetableSnapshot? newSnapshot)
    {
        if (newSnapshot == null) return Array.Empty<TimetableChangeRecord>();
        if (oldSnapshot == null)
        {
            // Initial import: no diff records needed
            return Array.Empty<TimetableChangeRecord>();
        }

        var oldMap = oldSnapshot.Periods.ToDictionary(p => p.Id, p => p);
        var newMap = newSnapshot.Periods.ToDictionary(p => p.Id, p => p);
        var changes = new List<TimetableChangeRecord>();

        // Check for modifications and additions
        foreach (var (newId, newPeriod) in newMap)
        {
            if (oldMap.TryGetValue(newId, out var oldPeriod))
            {
                // Check subject change
                if (!string.Equals(oldPeriod.Subject.Trim(), newPeriod.Subject.Trim(), StringComparison.OrdinalIgnoreCase))
                {
                    changes.Add(new TimetableChangeRecord
                    {
                        PeriodId = newId,
                        ChangeType = ChangeType.SubjectChanged,
                        OldSubject = oldPeriod.Subject,
                        NewSubject = newPeriod.Subject,
                        OldStartTime = oldPeriod.StartTime,
                        NewStartTime = newPeriod.StartTime,
                        OldEndTime = oldPeriod.EndTime,
                        NewEndTime = newPeriod.EndTime,
                        OldTeacher = oldPeriod.Details,
                        NewTeacher = newPeriod.Details
                    });
                }
                // Check time change
                else if (oldPeriod.StartTime != newPeriod.StartTime || oldPeriod.EndTime != newPeriod.EndTime)
                {
                    changes.Add(new TimetableChangeRecord
                    {
                        PeriodId = newId,
                        ChangeType = ChangeType.TimeChanged,
                        OldSubject = oldPeriod.Subject,
                        NewSubject = newPeriod.Subject,
                        OldStartTime = oldPeriod.StartTime,
                        NewStartTime = newPeriod.StartTime,
                        OldEndTime = oldPeriod.EndTime,
                        NewEndTime = newPeriod.EndTime,
                        OldTeacher = oldPeriod.Details,
                        NewTeacher = newPeriod.Details
                    });
                }
                // Check teacher/details change
                else if (!string.Equals(oldPeriod.Details.Trim(), newPeriod.Details.Trim(), StringComparison.OrdinalIgnoreCase))
                {
                    changes.Add(new TimetableChangeRecord
                    {
                        PeriodId = newId,
                        ChangeType = ChangeType.TeacherChanged,
                        OldSubject = oldPeriod.Subject,
                        NewSubject = newPeriod.Subject,
                        OldTeacher = oldPeriod.Details,
                        NewTeacher = newPeriod.Details
                    });
                }
            }
            else
            {
                // Added
                changes.Add(new TimetableChangeRecord
                {
                    PeriodId = newId,
                    ChangeType = ChangeType.Added,
                    NewSubject = newPeriod.Subject,
                    NewStartTime = newPeriod.StartTime,
                    NewEndTime = newPeriod.EndTime,
                    NewTeacher = newPeriod.Details
                });
            }
        }

        // Check for removals
        foreach (var (oldId, oldPeriod) in oldMap)
        {
            if (!newMap.ContainsKey(oldId))
            {
                changes.Add(new TimetableChangeRecord
                {
                    PeriodId = oldId,
                    ChangeType = ChangeType.Removed,
                    OldSubject = oldPeriod.Subject,
                    OldStartTime = oldPeriod.StartTime,
                    OldEndTime = oldPeriod.EndTime,
                    OldTeacher = oldPeriod.Details
                });
            }
        }

        return changes.AsReadOnly();
    }

    public TimetableSnapshot AnnotateSnapshot(TimetableSnapshot snapshot, IReadOnlyList<TimetableChangeRecord> changes)
    {
        var activeChangeMap = changes
            .Where(c => !c.IsAcknowledged)
            .GroupBy(c => c.PeriodId)
            .ToDictionary(g => g.Key, g => g.First());

        var updatedPeriods = new List<PeriodOccurrence>();
        foreach (var period in snapshot.Periods)
        {
            if (activeChangeMap.TryGetValue(period.Id, out var change))
            {
                updatedPeriods.Add(period with { ChangeRecord = change });
            }
            else
            {
                updatedPeriods.Add(period with { ChangeRecord = null });
            }
        }

        return snapshot with { Periods = updatedPeriods.AsReadOnly() };
    }
}
